<b>Привет!</b>

Меня зовут Харочкина Екатерина, и я буду проверять твой проект. Давай общаться на «ты», но если это неудобно, пожалуйста, сообщи. Моя основная цель — не указать на совершенные тобою ошибки, а поделиться своим опытом и помочь тебе совершенствоваться как профессионалу.

Ты подготовил замечательный проект, но есть аспекты, которые еще можно улучшить. Я буду использовать комментарии с цветовой разметкой. Пожалуйста, **не перемещай, не изменяй и не удаляй их**. Это поможет выполнить повторную проверку твоего проекта оперативнее.

<div class="alert alert-danger">
<b>Комментарий ревьюера №1</b> ❌
    
Таким комментарием выделены самые важные замечания. Без их отработки проект не будет принят.
</div>

<div class="alert alert-warning">
<b>Комментарий ревьюера №1</b> ⚠️
    
Таким комментарием выделены небольшие замечания и рекомендации, которые будет хорошо учесть.
</div>

<div class="alert alert-success">
<b>Комментарий ревьюера №1</b> ✅
    
Таким комментарием выделены удачные решения🙂
</div>

Свои комментарии ты можешь помечать синим цветом.

<div class="alert alert-info">
<b>Комментарий студента №1</b>
    
Так ты можешь оформить свой комментарий
</div>

Давай работать над проектом в диалоге: если ты **что-то меняешь** в проекте по моим рекомендациям — **пиши об этом**. После того, как ты все доработаешь, я смогу принять твой проект!) 

# Исследование надежности заемщиков


<div class="alert alert-success">
<b>Комментарий ревьюера №1</b> ✅
    
Хорошая практика начинать проект с контекста, описания задачи, возможного хода исследования, а также описания имеющихся данных. Можно тут также добавить содержание с гиперссылками. Про оформление работы можно почитать в материалах курса:
    
`Полезные инструкции для учёбы - Оформление проекта - Рекомендации по выполнению проектов`
</div>

Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [215]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [216]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [217]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


<div class="alert alert-success">
<b>Комментарий ревьюера №1</b> ✅
    
Разделы работы можно заканчивать промежуточными выводами
</div>

## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [218]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

<div class="alert alert-info">
<b>Комментарий студента №0</b>
    
Прикольное замечание! Принято, спасибо!
</div>

In [219]:
len(data.loc[(data['days_employed'].isna()) & (data['total_income'].isna())])

2174

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [220]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [221]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [222]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [223]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [224]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [225]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [226]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [227]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [228]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [229]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [230]:
data.duplicated().sum()

71

In [231]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [232]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [233]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [234]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [235]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [236]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

<div class="alert alert-danger">
<b>Комментарий ревьюера №1</b> ❌ Замечание по всему разделу ответов на вопросы 
    
В каждом вопросе нужно вывести сводную. Здесь можно воспользоваться `pivot_table()`.
    
Чтобы найти долю должников, нужно количество должников поделить на общее число заемщиков. Эту долю можно найти с помощью функции агрегации `mean`
    
Попробуй получить сводную следующего вида. В индексе столбец, который исследуем (в первом вопросе это количество детей), а столбцы: 'Всего кредитополучателей', 'Всего должников', 'Доля должников'. Учитывай, что использовать параметр columns необязательно, а в параметр aggfunc можно указывать несколько функций агрегации сразу.

Также могу посоветовать переименовать столбцы и написать функцию для построения сводных во всех вопросах.
</div>

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [237]:
print(data['children'].unique())

[1 0 3 2 4 5]


In [238]:
print('Число должников:', data[data['debt'] == 1]['purpose'].count())
print('Должники: детей от', data[data['debt'] == 1]['children'].min(), 'до', data[data['debt'] == 1]['children'].max())
print('Должники: медиана по числу детей:', data[data['debt'] == 1]['children'].median().astype(int))
print('Должники: среднее число детей:', data[data['debt'] == 1]['children'].mean())
print('Число бездолжников:', data[data['debt'] == 0]['purpose'].count())
print('Бездолжники: детей от', data[data['debt'] == 0]['children'].min(), 'до', data[data['debt'] == 0]['children'].max())
print('Бездолжники: медиана по числу детей:', data[data['debt'] == 0]['children'].median().astype(int))
print('Бездолжники: среднее число детей:', data[data['debt'] == 0]['children'].mean())

Число должников: 1732
Должники: детей от 0 до 4
Должники: медиана по числу детей: 0
Должники: среднее число детей: 0.5363741339491916
Число бездолжников: 19599
Бездолжники: детей от 0 до 5
Бездолжники: медиана по числу детей: 0
Бездолжники: среднее число детей: 0.4684932904740038


<div class="alert alert-info">
<b>Моя переделка по красному комментарию ревьюера №1</b>

**VVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVV**
</div> 

In [239]:
#data_pivot = data.pivot_table(index = ['children'], columns = 'debt', values = 'dob_years', aggfunc = {'count'})
data_pivot = data.pivot_table(index = ['children'],  values = 'debt', aggfunc = {'sum','count','mean'})
#data_pivot = data.pivot_table(index = ['debt','children'],values = 'dob_years',aggfunc = {'count'})
#data_pivot['1'].fillna(0)
#data_pivot['rate'] = data_pivot['1'] / (data_pivot['0'] + data_pivot['1'])
#data_pivot['total'] = data_pivot['sum'] / data_pivot['mean']
print(data_pivot)

            count      mean     sum
children                           
0         14091.0  0.075438  1063.0
1          4808.0  0.092346   444.0
2          2052.0  0.094542   194.0
3           330.0  0.081818    27.0
4            41.0  0.097561     4.0
5             9.0  0.000000     0.0


In [240]:
data_pivot = data.pivot_table(index = ['children'],  values = 'debt', aggfunc = {'sum','count','mean'})
display(data_pivot)

count      mean     sum
children                           
0         14091.0  0.075438  1063.0
1          4808.0  0.092346   444.0
2          2052.0  0.094542   194.0
3           330.0  0.081818    27.0
4            41.0  0.097561     4.0
5             9.0  0.000000     0.0

<div class="alert alert-success">
<b>Комментарий ревьюера №2</b> ✅ 
    
Для вывода датафрейма на экран можно использовать `display()` вместо `print()`
</div>

**Вывод:** Бездетные: 7.5% должников, 1 ребенок: 9.2% должников, 2 ребенка: 9.5% должников - статистика большая для этих категорий, видим, что наличие ребенка - скорее является фактором для невозврата кредита в срок, чем нет.

<div class="alert alert-danger">
<b>Комментарий ревьюера №1</b> ❌
    
Выявлять зависимость стоит, сравнивая самые большие, сопоставимые по размеру категории заемщиков. Какую зависимость можно выявить, сравнив категории с 0, 1 и 2 детьми?
</div>

<div class="alert alert-warning">
<b>Комментарий ревьюера №2</b> ⚠️
    
По людям с 0-2 детьми: чем больше детей, тем выше доля должников
</div>

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [26]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
print(data['family_status'].unique())
print()
print('debt == 1:',data[data['debt']==1].groupby('family_status')['children'].count())
print()
print('debt == 0:',data[data['debt']==0].groupby('family_status')['children'].count())
print()
print('Таблица, указывающая процент должников в каждой категории:')
for el in data['family_status'].unique():
    print(el, (100*(data[(data['debt'] == 1) & (data['family_status'] == el)]['children'].count() /
          data[data['family_status'] == el]['children'].count())).astype(int))

['женат / замужем' 'гражданский брак' 'вдовец / вдова' 'в разводе'
 'Не женат / не замужем']

debt == 1: family_status
Не женат / не замужем    273
в разводе                 84
вдовец / вдова            63
гражданский брак         385
женат / замужем          927
Name: children, dtype: int64

debt == 0: family_status
Не женат / не замужем     2523
в разводе                 1105
вдовец / вдова             888
гражданский брак          3749
женат / замужем          11334
Name: children, dtype: int64

Таблица, указывающая процент должников в каждой категории:
женат / замужем 7
гражданский брак 9
вдовец / вдова 6
в разводе 7
Не женат / не замужем 9


<div class="alert alert-info">
<b>Моя переделка по красному комментарию ревьюера №2</b>

**VVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVV**
</div> 

In [27]:
data_pivot2 = data.pivot_table(index = ['family_status'],  values = 'debt', aggfunc = {'sum','count','mean'})
print(data_pivot2)

                         count      mean    sum
family_status                                  
Не женат / не замужем   2796.0  0.097639  273.0
в разводе               1189.0  0.070648   84.0
вдовец / вдова           951.0  0.066246   63.0
гражданский брак        4134.0  0.093130  385.0
женат / замужем        12261.0  0.075606  927.0


**Вывод:** зависимость слабая. Разве что, люди побывавшие или пребывающие в документально подтвержденном браке более ответственны, нежели те, кто ни разу не вступал в законный брак.

<div class="alert alert-info">
<b>Связь с возрастом я бы не стал проводить, т.к. совместнопроживание становится все более модным, особенно для бездетных (а также для здоровых и не релоцирующихся) граждан.</b>
</div> 

<div class="alert alert-success">
<b>Комментарий ревьюера №1</b> ✅
    
Люди, побывавшие в официальном браке, наиболее надежны. Возможно, есть связь с возрастом
</div>

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [28]:
#print(data['total_income_category'].unique())
#print()
print('debt == 1:',data[data['debt']==1].groupby('total_income_category')['children'].count())
print()
print('debt == 0:',data[data['debt']==0].groupby('total_income_category')['children'].count())
print()
print('Таблица, указывающая процент должников в каждой категории:')
for el in data['total_income_category'].unique():
    print(el, (100*(data[(data['debt'] == 1) & (data['total_income_category'] == el)]['children'].count() /
          data[data['total_income_category'] == el]['children'].count())).astype(int))

debt == 1: total_income_category
A       2
B     354
C    1353
D      21
E       2
Name: children, dtype: int64

debt == 0: total_income_category
A       23
B     4660
C    14568
D      328
E       20
Name: children, dtype: int64

Таблица, указывающая процент должников в каждой категории:
B 7
C 8
D 6
E 9
A 8


<div class="alert alert-info">
<b>Моя переделка по красному комментарию ревьюера №3</b>

**VVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVV**
</div> 

In [29]:
data_pivot3 = data.pivot_table(index = ['total_income_category'],  values = 'debt', aggfunc = {'sum','count','mean'})
print(data_pivot3)

                         count      mean     sum
total_income_category                           
A                         25.0  0.080000     2.0
B                       5014.0  0.070602   354.0
C                      15921.0  0.084982  1353.0
D                        349.0  0.060172    21.0
E                         22.0  0.090909     2.0


**Вывод:** Статистика репрезентативна только для категорий дохода B-C. Для категорий A, D и E статистика довольно маленькая, не репрезентативная. Кат.B - 7.1% должников, кат.C - 8.5% должников. Чем больше доходов - тем лучше люди возвращают кредиты.

<div class="alert alert-danger">
<b>Комментарий ревьюера №1</b> ❌
    
Здесь лучше делать выводы по двум самым крупным категориям - B и С. По остальным категориям данных недостаточно для корректных выводов, статпогрешность большая
</div>

<div class="alert alert-success">
<b>Комментарий ревьюера №2</b> ✅
    
👍
</div>

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [30]:
print(data['purpose_category'].unique())
print()
print('debt == 1:',data[data['debt']==1].groupby('purpose_category')['children'].count())
print()
print('debt == 0:',data[data['debt']==0].groupby('purpose_category')['children'].count())
print()
print('Таблица, указывающая процент должников в каждой категории:')
for el in data['purpose_category'].unique():
    print(el, (100*(data[(data['debt'] == 1) & (data['purpose_category'] == el)]['children'].count() /
          data[data['purpose_category'] == el]['children'].count())).astype(int))# Ваш код будет здесь. Вы можете создавать новые ячейки.

['операции с недвижимостью' 'операции с автомобилем'
 'получение образования' 'проведение свадьбы']

debt == 1: purpose_category
операции с автомобилем      400
операции с недвижимостью    780
получение образования       369
проведение свадьбы          183
Name: children, dtype: int64

debt == 0: purpose_category
операции с автомобилем      3879
операции с недвижимостью    9971
получение образования       3619
проведение свадьбы          2130
Name: children, dtype: int64

Таблица, указывающая процент должников в каждой категории:
операции с недвижимостью 7
операции с автомобилем 9
получение образования 9
проведение свадьбы 7


<div class="alert alert-info">
<b>Моя переделка по красному комментарию ревьюера №4</b>

**VVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVV**
</div> 

In [31]:
data_pivot4 = data.pivot_table(index = ['purpose_category'],  values = 'debt', aggfunc = {'sum','count','mean'})
print(data_pivot4)

                            count      mean    sum
purpose_category                                  
операции с автомобилем     4279.0  0.093480  400.0
операции с недвижимостью  10751.0  0.072551  780.0
получение образования      3988.0  0.092528  369.0
проведение свадьбы         2313.0  0.079118  183.0


**Вывод:** Наиболее популярная цель кредита - недвижка, при этом тут, самый низкий процент образования долга - 7.3%, минимальное значение по категориям. Вероятно к этой задаче, делам с недвижимостью, люди подходят с самой большой ответственностью. Свадьбы - 7.9% надежности, самая маленький представитель выборки в нашей подборке, процент низкий, вероятно, т.к. много людей может помочь в этом вопросе - аж до 2-х семей! Наиболее рисковые - 9.3% - операции с авто и обучение. Обучение - вероятно не сразу люди могут найти работу после обучения для возврата займа. Авто - трудно сказать, скорее всего причина в том, что авто - тот еще пассив, который тратит зачастую много средств, ну и часто люди покупают авто дороже, чем следовало бы. Обычно говорят, что надо покупать авто на 0.5 годового дохода.

<div class="alert alert-danger">
<b>Комментарий ревьюера №1</b> ❌
    
Разница может быть значима для банка. Относительная разница составляет порядка 20%. Выдели, пожалуста, наиболее рискованные и наиболее надежные категории. Можно также подумать над причинами такого распределения
</div>

<div class="alert alert-success">
<b>Комментарий ревьюера №2</b> ✅
    
Хорошо
</div>

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

<div class="alert alert-info">
<b>Обновил:</b></div> 

*Ответ:*

Фактор пользователя-человека-клиента банка: не заполнил поля случайно или умышленно (лень или желание скрыть какие-то детали).

Виноват тех.специалист и данные не сохранились куда должны были или не выгрузились как/откуда нужно.

<div class="alert alert-danger">
<b>Комментарий ревьюера №1</b> ❌
    
Тут важно отметить человеческий и технический факторы. Клиенты могли не предоставить данные. Возможно, эту информацию не требуют в некоторых случаях. Проблемы могли возникнуть на этапе загрузки и на этапе выгрузки данных. Неправильный запрос, объединение таблиц также может быть причиной.
</div>

<div class="alert alert-warning">
<b>Комментарий ревьюера №1</b> ⚠️
    
Можно проверить, что пропуски в 'total_income' и 'days_employed' находятся на одних строках, например, так: 
    
    len(data.loc[(data['days_employed'].isna()) & (data['total_income'].isna())])
    
Значений 2174, что совпадает со значениями пропусков в обоих столбцах по отдельности.
</div>

<div class="alert alert-success">
<b>Комментарий ревьюера №2</b> ✅
    
Ок
</div>

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

<div class="alert alert-info">
<b>Vyvod is listed:</b></div> 

*Ответ:*

Если по каждой категории мы заполним пропущенные данные медианными (среди заполненных ранее), то не поменяется значение медианы в общей выборке выбранной категории (заполненные ранее + fillna медианой заполненных ранее). Потому что медиана - среднее значение в вариационной ряду и более репрезентативна для оценки средних количетсвенных величин, типа зп, температура и т.д. (Как в примере: 20-20-20-20-100 - mean == 36, median == 20). 

Если мы заполним пропуски **средними** значениями, а не медианными, то поменяется значение медианы. А ведь изначальное значение медианы для нас информативно, т.к. это настоящие данные! а не fillna

<div class="alert alert-danger">
<b>Комментарий ревьюера №1</b> ❌
    
Напиши, пожалуйста, почему мы заполнили пропуски медианой, а не средним, например, и почему нам важно сохранить значение медианы
</div>

<div class="alert alert-warning">
<b>Комментарий ревьюера №2</b> ⚠️
    
Среднее и медиана различаются, если в выборке присутствуют выбросы (значения, выделяющиеся из общей выборки). Медиана позволяет нивелировать их влияние.
</div>

### Шаг 4: общий вывод.

Было проведено исследование надежности заемщиков (НЗ). По анализу *имеющихся данных* можно сделать вывод о том, что нет зависимости между НЗ и уровнем дохода заемщика, нет зависимости между НЗ и количеством детей у заемщика, также нет зависимости между НЗ и целью получения кредита, как и нет зависимости между НЗ и семейным положением заемщика. Проценты находятся в пределах погрешности статистики.

<div class="alert alert-info">
<b>My new main Vyvod is listed:</b></div> 

# ГЛАВНЫЙ ВЫВОД ИССЛЕДОВАНИЯ

Было проведено исследование надежности заемщиков (НЗ).

Входные данные: 21331 записи о заемщиках.

Был проведен анализ НЗ в зависимости от

        числа детей заемщика,
        
        семейного положения заемщика,
        
        уровня дохода,
        
      и целью получения кредита.
      
Семейные люди более надежны, как заемщики. Люди с более высоким доходом более надежны, чем люди с меньшим доходом. К финансовым вопросам свадьбы и операциям с недвижимостью люди подходят более ответственно, чем к вопросам с авто и получением профессионального образования/следующей ступени образования. Чем меньше детей у заемщика - тем он более надежен.

**Выводы сделаны исключительно на выданных для этой задачи данных.**

<div class="alert alert-danger">
<b>Комментарий ревьюера №1</b> ❌
 
Выводы нужно будет исправить после изменений в проекте. В первую очередь обращают внимание на выводы - это отчет перед заказчиком. Здесь нужно описать какая работа над данными была проведена и каких результатов удалось добиться. Опиши все зависимости. Рекомендованная структура выводов следующая:
    
- абзац по предобработке данных, что было сделано
    
- выявленные зависимости просрочки от разных факторов в общем виде
    
- портреты надежного и рискованного заемщика
    
- рекомендации заказчику
</div>

<div class="alert alert-success">
<b>Комментарий ревьюера №2</b> ✅
    
Можно написать о том, что было сделано в предобработке. В качестве рекомендаций можно предложить другую категоризацию по доходам. Или исследовать другие признаки, рассчитать другие метрики (отношение дохода к сумме выплат и др.). В анкете сделать поля о доходе и стаже обязательными к заполнению, добавить автопроверку на формат вводимых пользователем данных
</div>

<div class="alert alert-info">
<b>Спасибо Екатерина! А можно "Катя"? Ты мой ревьюр навсегда или только на этот спринт?</b></div> 

<div class="alert alert-success">
<b>Комментарий ревьюера №2</b> ✅
    
Только на этот спринт. Есть вероятность, что будет мэтч в следующих проектах, но она мала.
</div>

<div style="border:solid purple 2px; padding: 20px"> 
<b> <font color='purple' >   
    
Ты хорошо постарался, но проект еще нуждается в доработке. Нужно построить сводные в каждом вопросе, проверить выводы, дополнить причины пропусков и ответ про заполнение пропусков. Доделывай и присылай на повторную проверку.
        
Жду твоих исправлений 😊
</font></b>
</div>

<div style="border:solid purple 2px; padding: 20px"> 
<b> <font color='purple' >   
Отлично! Критических замечаний больше нет. Проект принят! 👏 Поздравляю и желаю удачи!
</font></b>
</div>

In [1]:
a

NameError: name 'a' is not defined